### ISSUE TITLE
Kubernetes nodepool "Parameter Reference" documentation bug - Parameter types are incorrect

### ISSUE THREAD

**Issue by detmoo** [ionos-cloud/module-ansible#97]
## Description
On running the k8s nodepool task `ionoscloudsdk.ionoscloud.k8s_nodepool` with `state="update"`, we get "500 - Internal Server Error". Requesting help to diagnose the error or to realise any issues with our task inputs. _NB// this is possibly not a Bug, maybe only in so much as the behaviour of the module is possibly different from the docs_.

## Expected behavior
The following sequence:

(1) run k8s nodepool task `ionoscloudsdk.ionoscloud.k8s_nodepool` with `state="create"` and `node_count="1"`. **Expectation**: a nodepool is created with node_count="1". **Result**: _the nodepool is created correctly_.

(2) run k8s nodepool task `ionoscloudsdk.ionoscloud.k8s_nodepool` with `state="update"` and `node_count="2"`. **Expectation**: the above created nodepool is updated so that node_count="2". **Result**: _500 - Internal Server Error_.

## Environment

Ansible version:
```
ansible==5.8.0
ansible-core==2.12.6
ionoscloud==6.0.4
```
Module version: 
```
ionoscloudsdk.ionoscloud 6.2.0
```
OS: 
```
Ubuntu (Focal)
```

## How to Reproduce
Steps to reproduce the behaviour:
1. Following is the Invocation log for k8s nodepool task `ionoscloudsdk.ionoscloud.k8s_nodepool` with `state="create"` and `node_count="1"`:
```
2022-06-08T14:37:45.6618681Z     "invocation": {
2022-06-08T14:37:45.6619014Z         "module_args": {
2022-06-08T14:37:45.6619371Z             "annotations": null,
2022-06-08T14:37:45.6619666Z             "api_url": null,
2022-06-08T14:37:45.6619948Z             "auto_scaling": {
2022-06-08T14:37:45.6620235Z                 "max_node_count": "4",
2022-06-08T14:37:45.6620610Z                 "min_node_count": "1"
2022-06-08T14:37:45.6620884Z             },
2022-06-08T14:37:45.6621254Z             "availability_zone": "AUTO",
2022-06-08T14:37:45.6621643Z             "cores_count": "1",
2022-06-08T14:37:45.6621957Z             "cpu_family": "INTEL_SKYLAKE",
2022-06-08T14:37:45.6622628Z             "datacenter_id": "68e1a018-eaf5-40a4-b416-60d3ae6732ae",
2022-06-08T14:37:45.6623224Z             "k8s_cluster_id": "4dd2418f-fd65-4623-92f4-f89f135522c1",
2022-06-08T14:37:45.6623587Z             "k8s_version": "1.23.6",
2022-06-08T14:37:45.6623869Z             "labels": {
2022-06-08T14:37:45.6624155Z                 "status": "the_jungle_vip"
2022-06-08T14:37:45.6624434Z             },
2022-06-08T14:37:45.6624695Z             "lan_ids": null,
2022-06-08T14:37:45.6624987Z             "maintenance_window": {
2022-06-08T14:37:45.6626463Z                 "day_of_the_week": "Sunday",
2022-06-08T14:37:45.6626797Z                 "time": "03:45:00"
2022-06-08T14:37:45.6627064Z             },
2022-06-08T14:37:45.6627319Z             "node_count": 1,
2022-06-08T14:37:45.6627723Z             "nodepool_id": null,
2022-06-08T14:37:45.6630684Z             "nodepool_name": "king_of_the_swingers",
2022-06-08T14:37:45.6631884Z             "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
2022-06-08T14:37:45.6632229Z             "public_ips": null,
2022-06-08T14:37:45.6632517Z             "ram_size": "2048",
2022-06-08T14:37:45.6632805Z             "state": "present",
2022-06-08T14:37:45.6637384Z             "storage_size": "10",
2022-06-08T14:37:45.6637777Z             "storage_type": "SSD",
2022-06-08T14:37:45.6638136Z             "token": null,
2022-06-08T14:37:45.6639056Z             "username": "***",
2022-06-08T14:37:45.6639350Z             "wait": true,
2022-06-08T14:37:45.6639630Z             "wait_timeout": 600
2022-06-08T14:37:45.6639886Z         }
2022-06-08T14:37:45.6640117Z     } 
```
For the above invocation with state="create", the task performs as expected. Here is the corresponding log:
```
2022-06-08T14:37:45.6640371Z     "nodepool": {
2022-06-08T14:37:45.6641081Z         "href": "https://api.ionos.com/cloudapi/v6/k8s/4dd2418f-fd65-4623-92f4-f89f135522c1/nodepools/e93e937c-c089-4046-8bee-87eae77be578",
2022-06-08T14:37:45.6641916Z         "id": "e93e937c-c089-4046-8bee-87eae77be578",
2022-06-08T14:37:45.6654525Z         "metadata": {
2022-06-08T14:37:45.6655237Z             "created_by": "***",
2022-06-08T14:37:45.6655815Z             "created_by_user_id": "fe32a0f2-b40f-4332-846c-93dec3750045",
2022-06-08T14:37:45.6656612Z             "created_date": "2022-06-08T14:28:42+00:00",
2022-06-08T14:37:45.6657024Z             "etag": "3ad7bb1fcf47d74c51a2781d362b6e11",
2022-06-08T14:37:45.6657564Z             "last_modified_by": "***",
2022-06-08T14:37:45.6658273Z             "last_modified_by_user_id": "fe32a0f2-b40f-4332-846c-93dec3750045",
2022-06-08T14:37:45.6658872Z             "last_modified_date": "2022-06-08T14:28:42+00:00",
2022-06-08T14:37:45.6659236Z             "state": "DEPLOYING"
2022-06-08T14:37:45.6659500Z         },
2022-06-08T14:37:45.6659766Z         "properties": {
2022-06-08T14:37:45.6660055Z             "annotations": null,
2022-06-08T14:37:45.6660340Z             "auto_scaling": {
2022-06-08T14:37:45.6660645Z                 "max_node_count": 4,
2022-06-08T14:37:45.6660950Z                 "min_node_count": 1
2022-06-08T14:37:45.6661209Z             },
2022-06-08T14:37:45.6661604Z             "availability_zone": "AUTO",
2022-06-08T14:37:45.6661937Z             "available_upgrade_versions": null,
2022-06-08T14:37:45.6662248Z             "cores_count": 1,
2022-06-08T14:37:45.6662554Z             "cpu_family": "INTEL_SKYLAKE",
2022-06-08T14:37:45.6663101Z             "datacenter_id": "68e1a018-eaf5-40a4-b416-60d3ae6732ae",
2022-06-08T14:37:45.6663471Z             "k8s_version": "1.23.6",
2022-06-08T14:37:45.6663754Z             "labels": {
2022-06-08T14:37:45.6664049Z                 "status": "the_jungle_vip"
2022-06-08T14:37:45.6664315Z             },
2022-06-08T14:37:45.6664560Z             "lans": [],
2022-06-08T14:37:45.6664839Z             "maintenance_window": {
2022-06-08T14:37:45.6665154Z                 "day_of_the_week": "Sunday",
2022-06-08T14:37:45.6665460Z                 "time": "03:45:00Z"
2022-06-08T14:37:45.6665728Z             },
2022-06-08T14:37:45.6666015Z             "name": "king_of_the_swingers",
2022-06-08T14:37:45.6666316Z             "node_count": 1,
2022-06-08T14:37:45.6666597Z             "public_ips": null,
2022-06-08T14:37:45.6666873Z             "ram_size": 2048,
2022-06-08T14:37:45.6667162Z             "storage_size": 10,
2022-06-08T14:37:45.6667450Z             "storage_type": "SSD"
2022-06-08T14:37:45.6667705Z         },
2022-06-08T14:37:45.6667964Z         "type": "nodepool"
2022-06-08T14:37:45.6668214Z     },
2022-06-08T14:37:45.6668579Z     "results_file": "/home/***/.ansible_async/328711158290.399884",
2022-06-08T14:37:45.6668914Z     "started": 1,
2022-06-08T14:37:45.6669170Z     "stderr": "",
2022-06-08T14:37:45.6669439Z     "stderr_lines": [],
2022-06-08T14:37:45.6669698Z     "stdout": "",
2022-06-08T14:37:45.6669961Z     "stdout_lines": []
2022-06-08T14:37:45.6670206Z }
```
2. Following is the Invocation log for k8s nodepool task `ionoscloudsdk.ionoscloud.k8s_nodepool` with `state="update"` and `node_count="2"`. NB// The ``nodepool_id`` parameter matches the value output above. 
```
2022-06-08T15:04:37.1363032Z     "invocation": {
2022-06-08T15:04:37.1363415Z         "module_args": {
2022-06-08T15:04:37.1363809Z             "annotations": null,
2022-06-08T15:04:37.1364209Z             "api_url": null,
2022-06-08T15:04:37.1364604Z             "auto_scaling": {
2022-06-08T15:04:37.1365011Z                 "max_node_count": "4",
2022-06-08T15:04:37.1365422Z                 "min_node_count": "1"
2022-06-08T15:04:37.1365796Z             },
2022-06-08T15:04:37.1366191Z             "availability_zone": null,
2022-06-08T15:04:37.1366614Z             "cores_count": null,
2022-06-08T15:04:37.1367023Z             "cpu_family": null,
2022-06-08T15:04:37.1367432Z             "datacenter_id": null,
2022-06-08T15:04:37.1368135Z             "k8s_cluster_id": "4dd2418f-fd65-4623-92f4-f89f135522c1",
2022-06-08T15:04:37.1368639Z             "k8s_version": "1.23.6",
2022-06-08T15:04:37.1369039Z             "labels": {
2022-06-08T15:04:37.1369443Z                 "status": "the_jungle_vip"
2022-06-08T15:04:37.1369834Z             },
2022-06-08T15:04:37.1370202Z             "lan_ids": null,
2022-06-08T15:04:37.1370609Z             "maintenance_window": {
2022-06-08T15:04:37.1371037Z                 "day_of_the_week": "Sunday",
2022-06-08T15:04:37.1371476Z                 "time": "03:45:00"
2022-06-08T15:04:37.1371859Z             },
2022-06-08T15:04:37.1372227Z             "node_count": 2,
2022-06-08T15:04:37.1372877Z             "nodepool_id": "e93e937c-c089-4046-8bee-87eae77be578",
2022-06-08T15:04:37.1373372Z             "nodepool_name": null,
2022-06-08T15:04:37.1373840Z             "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
2022-06-08T15:04:37.1374301Z             "public_ips": null,
2022-06-08T15:04:37.1374708Z             "ram_size": null,
2022-06-08T15:04:37.1375106Z             "state": "update",
2022-06-08T15:04:37.1375513Z             "storage_size": null,
2022-06-08T15:04:37.1375918Z             "storage_type": null,
2022-06-08T15:04:37.1376316Z             "token": null,
2022-06-08T15:04:37.1376908Z             "username": "***",
2022-06-08T15:04:37.1377317Z             "wait": true,
2022-06-08T15:04:37.1377711Z             "wait_timeout": 600
2022-06-08T15:04:37.1378076Z         }
2022-06-08T15:04:37.1378413Z     }
```
For the above invocation with state="update" the task fails. Here is the corresponding log:
```
2022-06-08T15:04:37.1380608Z     "msg": "failed to update the nodepool: (500)\nReason: Internal Server Error\nHTTP response headers: HTTPHeaderDict({'Server': 'nginx', 'Date': 'Wed, 08 Jun 2022 15:04:20 GMT', 'Content-Type': 'application/json', 'Content-Length': '160', 'Connection': 'close', 'X-Frame-Options': 'SAMEORIGIN', 'Access-Control-Allow-Origin': '*', 'Access-Control-Allow-Headers': '*', 'Access-Control-Expose-Headers': '*', 'Access-Control-Allow-Methods': '*', 'Access-Control-Allow-Credentials': 'true'})\nHTTP response body: {\n  \"httpStatus\" : 500,\n  \"messages\" : [ {\n    \"errorCode\" : \"301\",\n    \"message\" : \"Oops! Something went very wrong. Please contact the administrator\"\n  } ]\n}\n\n"
```

## Additional Notes

Observe which parameters are set to ``null`` in the invocation logs. We have tried to ensure ``null`` is passed for at least any parameters that are not explicitly handled by the underlying python function for the ``"create"`` and `"update"` cases. In the `"update"` case, we are looking at parameters handled by function `update_k8s_cluster_nodepool()` in [k8s_nodepool.py](https://github.com/ionos-cloud/module-ansible/blob/master/plugins/modules/k8s_nodepool.py). For example, ``cores_count`` is set to null. We believe this approach is also consistent with the Examples and Parameter Reference in the [docs](https://docs.ionos.com/ansible/api/managed-kubernetes/k8snodepool).

## References
#54

**Comment by avirtopeanu-ionos** [comment]
Hello @detmoo, I can't seem to reproduce the error. Is this still happening? It seems like the call is passed to the API correctly by the ansible module, and fails. Can you try running the playbook again?

**Comment by detmoo** [comment]
Hello @avirtopeanu-ionos I am still experiencing the same issue. I agree the call is passed OK to the API but it is not clear what is causing the API to issue the 500 error and therefore, based on the evidence I have, it is not possible to rule out a flaw in the module on transposing parameters from the Ansible playbook to the underlying API call.

Since you have been unable to recreate the error, so please can you share the invocation logs generated on successfully processing your 'update' playbook so that I can reconcile against what I have provided above?

For your information, I have attempted to validate the parameter transpose by tracing the lineage from the parameter interface implemented by the Ansible module through to the parameter interface implemented by the underlying Python SDK (i.e. [KubernetesNodepoolPropertiesForPut](https://docs.ionos.com/python-sdk/models/kubernetesnodepoolpropertiesforput) but I have not found any issues. Should the problem be connected to parameters, to me it might be reasonable to expect successfully parsing the parameter interface in Ansible to cause deterministically successful parsing of the parameter interface for the underlying API call.

**Comment by avirtopeanu-ionos** [comment]
Running the following playbook is fine for me

```
- name: Detmoo issue97
  hosts: localhost
  connection: local
  gather_facts: false

  vars_files:
    - ../vars.yml

  tasks:
    - name: Create datacenter
      datacenter:
        name: "{{ datacenter }}"
        description: "{{ description }}"
        location: "{{ location }}"
      register: datacenter_response

    - name: Create k8s cluster
      k8s_cluster:
        cluster_name: "{{ cluster_name }}"
        maintenance_window:
          day_of_the_week: "Wednesday"
          time: "12:02:00"
        wait: true
      register: cluster

    - name: Create k8s cluster nodepool
      k8s_nodepool:
        nodepool_name: "{{ nodepool_name }}"
        k8s_cluster_id: "{{ cluster.cluster.id }}"
        datacenter_id: "{{ datacenter_response.datacenter.id }}"
        node_count: 1
        cpu_family: "INTEL_SKYLAKE"
        cores_count: "1"
        ram_size: "2048"
        storage_type: null
        storage_size: null
        availability_zone: "AUTO"
        storage_type: "SSD"
        storage_size: "10"
        public_ips: null
        labels:
          status: "the_jungle_vip"
        annotations: null
        auto_scaling:
          max_node_count: 4
          min_node_count: 1
        wait: true
        state: present
      register: result

    - name: Debug - Show nodepool
      debug:
         msg: "{{ result.nodepool }}"

    - name: Update k8s cluster nodepool
      k8s_nodepool:
        k8s_cluster_id: "{{ cluster.cluster.id }}"
        datacenter_id: "{{ datacenter_response.datacenter.id }}"
        nodepool_id: "{{ result.nodepool.id }}"
        node_count: 2
        cpu_family: null
        cores_count: null
        ram_size: null
        # cpu_family: "INTEL_SKYLAKE"
        # cores_count: "1"
        # ram_size: "2048"
        availability_zone: null
        storage_type: null
        storage_size: null
        # availability_zone: "AUTO"
        # storage_type: "HDD"
        # storage_size: "100"
        public_ips: null
        # public_ips: "{{ ipblock_response_update.ipblock.properties.ips }}"
        labels:
          status: "the_jungle_vip"
        annotations: null
        auto_scaling:
          max_node_count: 4
          min_node_count: 1
        wait: true
        state: update
```

In the meantime I've created a ticket for our API team to investigate the underlying API calls, and I'll return with a comment here once I have more updates

**Comment by detmoo** [comment]
With some modifications* it has been possible to run the Playbook that you provided above. Using this outcome I anticipate to deduce a malformed parameter in my original playbook invocation. Nevertheless, there are still some issues with the results of the above playbook:-

(*_For your information_ - **Q.** What modifications were necessary to complete the playbook run? **A.** **(i)** running on an Azure hosted agent it is necessary to apply `async` and `poll` parameters to the playbook - otherwise the Ansible runtime hangs on the first task, even though it will successfully create but not update the nodepool; **(ii)** we adopted the fully-qualified name for referencing the Ionoscloud Tasks.)

Using the necessary modifications for my environment I can now see the playbook running to completion - but the nodepool still does not conform to the updated/declared specification. Following is (a) from the Ansible logs, observe we are declaring the nodepool shall be updated to have **2** nodes (updated from 1 node) (b) response from `ionoscloud.KubernetesApi().k8s_nodepools_get()`, observe the resource is in a terminal state ("ACTIVE") but with `'node_count': 1` (we expect 2, given the completion of the playbook). Please can you examine results from your running the playbook and advise if the outcome I am seeing is recreated in your environment?

**(a) Ansible logs - declares update of node count so that `'node_count': 2`**

```
2022-07-11T15:56:42.3020337Z changed: [localhost] => {
2022-07-11T15:56:42.3020695Z     "action": "update",
2022-07-11T15:56:42.3025168Z     "ansible_job_id": "865076800348.17353",
2022-07-11T15:56:42.3025541Z     "changed": true,
2022-07-11T15:56:42.3025834Z     "finished": 1,
2022-07-11T15:56:42.3026104Z     "invocation": {
2022-07-11T15:56:42.3026396Z         "module_args": {
2022-07-11T15:56:42.3026706Z             "annotations": null,
2022-07-11T15:56:42.3027015Z             "api_url": null,
2022-07-11T15:56:42.3027318Z             "auto_scaling": {
2022-07-11T15:56:42.3027836Z                 "max_node_count": 4,
2022-07-11T15:56:42.3028159Z                 "min_node_count": 1
2022-07-11T15:56:42.3028440Z             },
2022-07-11T15:56:42.3030331Z             "availability_zone": null,
2022-07-11T15:56:42.3030777Z             "cores_count": null,
2022-07-11T15:56:42.3031170Z             "cpu_family": null,
2022-07-11T15:56:42.3036166Z             "datacenter_id": "54f7da8c-1ed6-4651-bef7-36ee648edc5d",
2022-07-11T15:56:42.3039320Z             "k8s_cluster_id": "83493f82-a318-4de1-972e-99495740ea88",
2022-07-11T15:56:42.3041651Z             "k8s_version": null,
2022-07-11T15:56:42.3042184Z             "labels": {
2022-07-11T15:56:42.3044164Z                 "status": "the_jungle_vip"
2022-07-11T15:56:42.3044663Z             },
2022-07-11T15:56:42.3044962Z             "lan_ids": null,
2022-07-11T15:56:42.3045484Z             "maintenance_window": null,
2022-07-11T15:56:42.3045996Z             "node_count": 2,
2022-07-11T15:56:42.3047010Z             "nodepool_id": "538d5abc-82e1-4cb3-b610-b56d8ce3c6a8",
2022-07-11T15:56:42.3047701Z             "nodepool_name": null,
2022-07-11T15:56:42.3048804Z             "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
2022-07-11T15:56:42.3049665Z             "public_ips": null,
2022-07-11T15:56:42.3050528Z             "ram_size": null,
2022-07-11T15:56:42.3050987Z             "state": "update",
2022-07-11T15:56:42.3055208Z             "storage_size": null,
2022-07-11T15:56:42.3055572Z             "storage_type": null,
2022-07-11T15:56:42.3056038Z             "token": null,
2022-07-11T15:56:42.3057481Z             "username": "***",
2022-07-11T15:56:42.3057857Z             "wait": true,
2022-07-11T15:56:42.3058201Z             "wait_timeout": 600
2022-07-11T15:56:42.3058530Z         }
2022-07-11T15:56:42.3058816Z     },
2022-07-11T15:56:42.3059127Z     "nodepool": {
2022-07-11T15:56:42.3060544Z         "href": "https://api.ionos.com/cloudapi/v6/k8s/83493f82-a318-4de1-972e-99495740ea88/nodepools/538d5abc-82e1-4cb3-b610-b56d8ce3c6a8",
2022-07-11T15:56:42.3061985Z         "id": "538d5abc-82e1-4cb3-b610-b56d8ce3c6a8",
2022-07-11T15:56:42.3064090Z         "metadata": {
2022-07-11T15:56:42.3066756Z             "created_by": "***",
2022-07-11T15:56:42.3069551Z             "created_by_user_id": "fe32a0f2-b40f-4332-846c-93dec3750045",
2022-07-11T15:56:42.3072144Z             "created_date": "2022-07-11T15:41:17+00:00",
2022-07-11T15:56:42.3074347Z             "etag": "b787c06ddb5dcaddcf2e9463faa31c9e",
2022-07-11T15:56:42.3075860Z             "last_modified_by": "***",
2022-07-11T15:56:42.3078097Z             "last_modified_by_user_id": "fe32a0f2-b40f-4332-846c-93dec3750045",
2022-07-11T15:56:42.3080979Z             "last_modified_date": "2022-07-11T15:51:02+00:00",
2022-07-11T15:56:42.3081408Z             "state": "UPDATING"
2022-07-11T15:56:42.3082139Z         },
2022-07-11T15:56:42.3082684Z         "properties": {
2022-07-11T15:56:42.3083018Z             "annotations": {},
2022-07-11T15:56:42.3083350Z             "auto_scaling": {
2022-07-11T15:56:42.3084043Z                 "max_node_count": 4,
2022-07-11T15:56:42.3084577Z                 "min_node_count": 1
2022-07-11T15:56:42.3085248Z             },
2022-07-11T15:56:42.3085863Z             "availability_zone": "AUTO",
2022-07-11T15:56:42.3086301Z             "available_upgrade_versions": [
2022-07-11T15:56:42.3086690Z                 "1.22.9",
2022-07-11T15:56:42.3087035Z                 "1.22.11"
2022-07-11T15:56:42.3088748Z             ],
2022-07-11T15:56:42.3089217Z             "cores_count": 1,
2022-07-11T15:56:42.3091035Z             "cpu_family": "INTEL_SKYLAKE",
2022-07-11T15:56:42.3091873Z             "datacenter_id": "54f7da8c-1ed6-4651-bef7-36ee648edc5d",
2022-07-11T15:56:42.3094220Z             "k8s_version": "1.22.8",
2022-07-11T15:56:42.3094561Z             "labels": {
2022-07-11T15:56:42.3094893Z                 "status": "the_jungle_vip"
2022-07-11T15:56:42.3095459Z             },
2022-07-11T15:56:42.3095751Z             "lans": [],
2022-07-11T15:56:42.3096607Z             "maintenance_window": {
2022-07-11T15:56:42.3096966Z                 "day_of_the_week": "Sunday",
2022-07-11T15:56:42.3097323Z                 "time": "13:02:17Z"
2022-07-11T15:56:42.3097612Z             },
2022-07-11T15:56:42.3098352Z             "name": "debugNodepool",
2022-07-11T15:56:42.3098713Z             "node_count": 2,
2022-07-11T15:56:42.3099149Z             "public_ips": null,
2022-07-11T15:56:42.3099582Z             "ram_size": 2048,
2022-07-11T15:56:42.3101547Z             "storage_size": 10,
2022-07-11T15:56:42.3102117Z             "storage_type": "SSD"
2022-07-11T15:56:42.3103906Z         },
2022-07-11T15:56:42.3104233Z         "type": "nodepool"
2022-07-11T15:56:42.3105517Z     },
2022-07-11T15:56:42.3105980Z     "results_file": "/home/***/.ansible_async/865076800348.17353",
2022-07-11T15:56:42.3106641Z     "started": 1,
2022-07-11T15:56:42.3106964Z     "stderr": "",
2022-07-11T15:56:42.3107287Z     "stderr_lines": [],
2022-07-11T15:56:42.3108780Z     "stdout": "",
2022-07-11T15:56:42.3110644Z     "stdout_lines": []
2022-07-11T15:56:42.3113624Z }
```

**(b) `ionoscloud.KubernetesApi().k8s_nodepools_get()` shows `'node_count': 1` (not 2)**

```
{'href': 'https://api.ionos.com/cloudapi/v6/k8s/83493f82-a318-4de1-972e-99495740ea88/nodepools',
 'id': '83493f82-a318-4de1-972e-99495740ea88/nodepools',
 'items': [{'href': 'https://api.ionos.com/cloudapi/v6/k8s/83493f82-a318-4de1-972e-99495740ea88/nodepools/538d5abc-82e1-4cb3-b610-b56d8ce3c6a8',
            'id': '538d5abc-82e1-4cb3-b610-b56d8ce3c6a8',
            'metadata': {'created_by': 'oliver@bow-data.com',
                         'created_by_user_id': 'fe32a0f2-b40f-4332-846c-93dec3750045',
                         'created_date': datetime.datetime(2022, 7, 11, 15, 41, 17, tzinfo=tzutc()),
                         'etag': '3406c3baa51c6333596989b239ef19df',
                         'last_modified_by': 'oliver@bow-data.com',
                         'last_modified_by_user_id': 'fe32a0f2-b40f-4332-846c-93dec3750045',
                         'last_modified_date': datetime.datetime(2022, 7, 11, 15, 51, 2, tzinfo=tzutc()),
                         'state': 'ACTIVE'},
            'properties': {'annotations': {},
                           'auto_scaling': {'max_node_count': 4,
                                            'min_node_count': 1},
                           'availability_zone': 'AUTO',
                           'available_upgrade_versions': ['1.22.9', '1.22.11'],
                           'cores_count': 1,
                           'cpu_family': 'INTEL_SKYLAKE',
                           'datacenter_id': '54f7da8c-1ed6-4651-bef7-36ee648edc5d',
                           'k8s_version': '1.22.8',
                           'labels': {'status': 'the_jungle_vip'},
                           'lans': [],
                           'maintenance_window': {'day_of_the_week': 'Sunday',
                                                  'time': '13:02:17Z'},
                           'name': 'debugNodepool',
                           'node_count': 1,
                           'public_ips': None,
                           'ram_size': 2048,
                           'storage_size': 10,
                           'storage_type': 'SSD'},
            'type': 'nodepool'}],
 'type': 'collection'}
```

**Comment by detmoo** [comment]
@avirtopeanu-ionos FYI - I am still working on the evidence provided in the previous post. Possibly we can now close this issue but a new one could be opened concerning what is described [above](https://github.com/ionos-cloud/module-ansible/issues/97#issuecomment-1180627867). I will update or close issues shortly. Appreciate any more feedback from you in the mean time.

**Comment by avirtopeanu-ionos** [comment]
What's happening is that the nodepool is updated, node_count is set to two, until the autoscaler kicks in (see provided arguments: auto_scaling': {'max_node_count': 4, min_node_count': 1}, and then the autoscaler scales it back down to 1, so this is pretty much intended behaivour.

you can either remove the auto_scaling parameter (if you don't intend this to happen) or to raise its min_node_count to 2

**Comment by detmoo** [comment]
Hi, thank you for the explanation regarding the auto-scaling. 

Regarding the original issue. We have indeed deduced there was a malformed parameter in the `update` task. Namely, we were passing strings for the values of `min_node_count` and `max_node_count`. Once we change these fields to integers, the `update` task runs to completion. Nevertheless, the expectation of integers in the `update` case is obscure since it looks like the `create` task works fine with strings for these fields.

As a general point, maybe there is some inconsistency with the use of integers (`node_count`, `min_node_count`, `max_node_count`) and strings (`ram_size`, `storage_size`, `cores_count`) that could be clearer. FYI - as a workaround solution to #54, I have created a generic jinja2 template for all operations on nodepools (i.e.  for `state in {"create" | "update" | "delete"}`) with some of the parameter handling performed via jinja2 filters. And in this scenario it is *arguably* preferable to accept strings where possible, to avoid [this issue with jinja2](https://github.com/ansible/ansible/issues/30366).

I will appreciate any updates on these points regarding integers/strings but unless you have further feedback I shall close this issue. Thank you very much.

**Comment by detmoo** [comment]
FYI - check this page from the [ansible module docs](https://docs.ionos.com/ansible/api/managed-kubernetes/k8snodepool)and observe "**Example Syntax**". In `Create k8s cluster nodepool` the integer-like data are represented as a mixture of `int` and `str` including `node_count` (`int`). In `Update k8s cluster nodepool` we have `node_count` (`str`). Then obsever "**Parameter Reference**" and you have `node_count` (`int`).

**Comment by detmoo** [comment]
@avirtopeanu-ionos further to the comments above regarding need to wrangle the autoscaling parameters `min_node_count` and `max_node_count` as integers (_cf. it is necessary to pass integers to the `update` task even though the `create` task accepts strings_) - in fact [it seems to be not possible to wrangle the required integers in Ansible jinja2 templates](https://stackoverflow.com/questions/69109062/cant-set-ansible-fact-as-integer). Therefore, I kindly and respectfully request that you make allowance for this when you refactor the module in respect of the issues I raise [here](https://github.com/ionos-cloud/module-ansible/issues/97#issuecomment-1183388687) and [here](https://github.com/ionos-cloud/module-ansible/issues/97#issuecomment-1183494329). 

For me, you can close this issue once you have assimilated. Let me know if there any questions.

**Comment by avirtopeanu-ionos** [comment]
Hi, I will re-open this issue, and change its title accordingly. I agree that there is a bug within the documentation regarding the parameter types. 

Regarding the declarative tasks, the jinja2 templating is a smart workaround, but a long-term goal is that Ansible should guarantee the properties of resources (which it currently doesn't for `create` - and #54 is a testament to this)

We really appreciate the feedback, and please keep in mind that your issues are very helpful to this repo's development.

**Comment by rmocanu-ionos** [comment]
This should be solved in 7.4.1, please let us know if there are other issues.
